server/aws-lsp-codewhisperer/src/language-server/securityScan/securityScanDiagnosticsProvider.ts

```diff
@@ -0,0 +1,130 @@
+import { Logging, Lsp } from '@aws/language-server-runtimes/out/features'
+import { Diagnostic, Hover, Position, Range, TextDocumentContentChangeEvent } from 'vscode-languageserver'
+import { URI } from 'vscode-uri'
+import { AggregatedCodeScanIssue, CodeScanIssue } from './types'
+
+class SecurityScanDiagnosticsProvider {
+    private diagnostics: Map<string, Diagnostic[]>
+    private lsp: Lsp
+    private logging: Logging
+
+    constructor(lsp: Lsp, logging: Logging) {
+        this.diagnostics = new Map()
+        this.lsp = lsp
+        this.logging = logging
+    }
+
+    async createDiagnostics(findings: AggregatedCodeScanIssue[]) {
+        for (const finding of findings) {
+            const path = URI.file(finding.filePath).path
+            const diagnostics = finding.issues.map(issue => this.mapScanIssueToDiagnostics(issue))
+            if (!this.diagnostics.has(path)) {
+                this.diagnostics.set(path, diagnostics)
+            }
+            this.diagnostics.set(path, [...(this.diagnostics.get(finding.filePath) || []), ...diagnostics])
+            await this.publishDiagnostics(finding.filePath, diagnostics)
+        }
+    }
+
+    mapScanIssueToDiagnostics(issue: CodeScanIssue): Diagnostic {
+        return Diagnostic.create(
+            this.createDiagnosticsRange(issue.startLine, issue.endLine),
+            `${issue.title} - ${issue.description.text}`,
+            2,
+            issue.relatedVulnerabilities.join(','),
+            'Detected by CodeWhisperer'
+        )
+    }
+
+    async validateDiagnostics(uri: string, e: TextDocumentContentChangeEvent) {
+        const path = URI.parse(uri).path
+        const currentDiagnostics = this.diagnostics.get(path)
+        if (!currentDiagnostics) {
+            return
+        }
+        const nextDiagnostics: Diagnostic[] = currentDiagnostics.map(diagnostic => {
+            if ((e as any).range.start.line > diagnostic.range.end.line) {
```

Comment:
How do we ensure the TextDocumentContentChangeEvent we receive is not of type {text:string} here? Is this IDE specific behaviour?

Reply:
seems like if there is only text, that implies the range is the entire document. Will make some updates to cover this case. Thanks for catching

Reply:
If this is IDE specific and we don't want to handle cases where server receives the full updated text document contents rather than a range, we can maybe use a typecheck function like this:

```
function isTextDocumentContentChangeEventWithRange(
    e: TextDocumentContentChangeEvent
): e is TextDocumentContentChangeEvent & { range: Range } {
    return 'range' in e;
}
```

so the code would start with a type check and we would returns the original diagnostic if the change event doesn't contain the range:

```
if (!isTextDocumentContentChangeEventWithRange(e)) {
   return diagnostic
}

if (e.range.start.line > diagnostic.range.end.line) {...}
else{...}
...
```

Reply:
```
seems like if there is only text, that implies the range is the entire document. Will make some updates to cover this case. Thanks for catching

```

That makes it easier then, nice

Reply:
updated